### ISSUE TITLE
Misdetected cycle on combination of `finalizedBy` and `mustRunAfter` constraints

### ISSUE THREAD

**Issue by bamboo** [gradle/gradle#20800]
The following example reconstructed from a real project will get Gradle to fail due to a _Misdetected cycle between :removeContainer and :dockerTest. Help us by reporting this to https://github.com/gradle/gradle/issues/2293_.

```
            tasks {
                register("dockerTest") {
                    dependsOn("dockerUp")     // dependsOn createContainer mustRunAfter removeContainer
                    finalizedBy("dockerStop") // dependsOn removeContainer
                }

                register("dockerUp") {
                    dependsOn("createContainer")
                }

                register("dockerStop") {
                    dependsOn("removeContainer")
                }

                register("createContainer") {
                    mustRunAfter("removeContainer")
                }

                register("removeContainer") {
                }
            }
```

## Expected Behavior
A dependency of a finalizer must not run until it is known to be needed by something else that should run. So if the dependency is only required by a finalizer, then it should not start until the finalizer is ready to start (ie the finalized task has completed). But if the dependency is also required by some node reachable from the command-line tasks, then it should start as soon as its dependencies are ready. Or if the dependency is required by multiple finalizers, then it should not start until one of those finalizer is ready to start.

**Comment by bamboo** [comment]
Superseded by #20910.
